### DISCUSSION TITLE
PerUserFilter

### DISCUSSION THREAD

**Discussion by ideas2appsaustralia** [Azure/azure-mobile-apps#426] [Q&A] [answered]
Adrian can you pint me in the right direction. I want to download from remote table on server a subset from a table not based on Id, UpdatedAt etc but based on one of my own properties which is a non-generic EntityTableData field and so had used your Extension. 

`public static class TodoItemExtensions
    {
        public static IQueryable<TodoItem> PerUserFilter(this IQueryable<TodoItem> query, string userid)
        {
            return query.Where(item => item.UserId.Equals(userid));
        }
    }`

and thenin table controller used  Query()

`public IQueryable<TodoItem> GetAllTodoItems()
        {
            return Query().PerUserFilter(UserId);
        }`

Has Query() been replaced by QueryAsync() and if so how do I construct the extension so the filter works ?  IDomainManager was also used in conjunction with Query(). Is this required perhaps?

**Comment by fileman** [comment, marked as answer]
Have you tried to use [IAccessControlProvider](https://docs.microsoft.com/en-us/azure/developer/mobile-apps/azure-mobile-apps/howto/server/dotnet-core#configure-access-permissions)?

**Reply by ideas2appsaustralia** [reply]
Hi Emanuele

I probably need a bit more help than usual as I am not a professional developer

I am not sure about IAccessControlProvider as it reads as if it is designed for controlling user access or are you referring to the GetDataView ? Do you mean I could modify GetDataView to filter on any other field and it will return the filtered subset?

`public Expression<Func<T,bool>> GetDataView()
    {
      return (UserId == null)
        ? _ => false
        : model => model.UserId == UserId;
    }`

I just had a look in the TableController class and I can now see this line in the QUeryAsync method. This is also similar to what I used in the old framework which was returning an IQueryable, but was using IDomainManager.

`IQueryable<TEntity> queryable = Repository.AsQueryable().ApplyDataView(AccessControlProvider.GetDataView()).ApplyDeletedView(base.Request, Options.EnableSoftDelete);`

so am I right that I modify GetDataView and just change model.UserId==UserId to be whichever field I want to filter on?

anyway I will play around with it

thanks
Scott

**Reply by adrianhall** [reply]
The HOWTO for the server shows this EXACT scenario!

**Reply by ideas2appsaustralia** [reply]
Adrian sorry are you referring to the IAccessControlProvider fileman prposed? I cant see any other code on those  Server HOWTO pages that look like what Im after or are you referring to this [https://docs.microsoft.com/en-us/azure/developer/mobile-apps/azure-mobile-apps/howto/client/dotnet#filtering-data](url)

can you provide a link if its not either of the above?

**Reply by adrianhall** [reply]
I was actually meaning in the server: https://docs.microsoft.com/en-us/azure/developer/mobile-apps/azure-mobile-apps/howto/server/dotnet-core#configure-access-permissions

You can restrict the data view in two places:

1. On the server - where you (as the developer) only allow the user to see a portion of the dataset) - in the link, the user can only see their own records.
2. On the client - the user is allowed to see all records, but you only want to download a portion for offline usage.

Based on your original question (where you are trying to adjust the table controller), I believe it's #1 you want.

**Reply by ideas2appsaustralia** [reply]
I want to download from the server a subset of a table based on a filter value the user selects. In this instance the user can access all records depending on what they choose, but I want to avoid downloading the whole table to the client then filtering. I want the filtering to occur server side to reduce number of records being downloaded from server.

So maybe its #2? Anyway can you link me to code example for #2 and then I have both.

**Reply by ideas2appsaustralia** [reply]
there is the example for getting 1 record using the tables Id field. I want to construct a filter eg. where this == that & where that != this etc where I can filter on any field in the table I want.

**Reply by fileman** [reply]
no, I think you need the option 1. 
you can copy the code in docs and maybe only minimal changes are required by your solution to have server side filter data by userId,

1. implement PrivateAccessControlProvider (copy from docs page)
2. use it in controller as shown in docs 
`AccessControlProvider = new PrivateAccessControlProvider<Model>(accessor);`

With those when your client app call server query controller method che AccessControlProvider filter table and return only item with same userid of authenticated user.

**Comment by ideas2appsaustralia** [comment]
Emanuele I dont want to filter by UserId

if my table has a field called MyColor then I want to filter using for 
example .Where model.MyColor == "blue"

i just want this filter to occur server side so I dont download all 
records to client device.

is that different to what AccessControlProvider is meant for?

thanks
scott

On 7/07/2022 7:43 am, Emanuele Filardo wrote:
>
> no, I think you need the option 1.
> you can copy the code in docs and maybe only minimal changes are 
> required by your solution to have server side filter data by userId,
>
>  1. implement PrivateAccessControlProvider (copy from docs page)
>  2. use it in controller as shown in docs
>     |AccessControlProvider = new
>     PrivateAccessControlProvider<Model>(accessor);|
>
> With those when your client app call server query controller method 
> che AccessControlProvider filter table and return only item with same 
> userid of authenticated user.
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/Azure/azure-mobile-apps/discussions/426#discussioncomment-3095553>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AXS3BMRHKMMKX3ARR4SH2A3VSX4Y7ANCNFSM52PXMZ4Q>.
> You are receiving this because you authored the thread.Message ID: 
> ***@***.***>
>

-- 
Regards Scott

**Reply by fileman** [reply]
your first messagge is about userfilter ... to filter table based on property, you have to use simple [filter](https://docs.microsoft.com/en-us/azure/developer/mobile-apps/azure-mobile-apps/howto/client/dotnet#filtering-data)

**Reply by ideas2appsaustralia** [reply]
Hi

yes it says PerUserFilter but it was an extension on the table controller that allowed you to build an odata query on the server side if I recall accurately. It came from Adrians iBook using the old windows.mobile framework.

anyway regarding the simple query the reason I was unsure about this is that it says that with first query it occurs on server then second query with the x.Title example occurs on CLient so I understood this to mean the whole table is first sent to the mobile client then I query on the other fields so still not what I am after.

but maybe I am wrong

scott

**Reply by adrianhall** [reply]
When you do the following:

``` csharp
var items = await _table.Where(m => m.Field == "some value").ToListAsync();
```

The query happens on the server side.  Only the matching records are returned.  When you do the following:

``` csharp
var items = (await _table.ToListAsync()).Where(m => m.Field == "some value").ToList();
```

The entire contents of the table are returned and filtered on the client side.  Now, if you wanted to store a user-defined subset of records on the client offline cache, you can do something like this:

``` csharp
var query = _table.CreateQuery().Where(m => m.Field == "some value");
await _table.PullItemsAsync(query);
```

This will only pull items that match the query.  You can then use the offline table query facilities:

``` chsarp
var items = _table.Where(m => m.Field == "some other value").ToListAsync(); // No items because they didn't get pulled
```

Hope this helps!

**Reply by ideas2appsaustralia** [reply]
thanks Adrian and Emanuele. Its getting clearer.

In my case I am using CloudTable<T> service where T : DatasyncClientData and so the name of the table is unknown until runtime as its a generic service. From your iBook Adrian. Whenever the service is called any of the many DatasyncClientData tables can be used so its kinda hard to work out the logic so that when a particular table is being used the right query is constructed since its not known until the Service is used. The PerUserFilter extension operated on the TableController so this problem did not occur as the table type was known already by the controller so it recognized the fields from the model.

Adrian this type of use case that you had in your iBook is much more realistic a scenario than the one single ToDo table in the example. At least for a relative beginner. I know this is not really an exercise in learning c# .net coding but for some idiots like me its invaluable!

**Reply by ideas2appsaustralia** [reply]
Adrian you mentioned in your youtube debut List Items that had query filtering (OData) as one of the CRUD actions. This sounded to me like what I need to solve my above problem? Where is that in the Library?

**Reply by adrianhall** [reply]
It's in the client and service - in fact, it's a big part of the whole service.  Check out the TableController<T> code for the server side implementation.

**Reply by adrianhall** [reply]
And, to implement a per-user filter, there is an actual example of this in the HOW TO here: https://docs.microsoft.com/en-us/azure/developer/mobile-apps/azure-mobile-apps/howto/server/dotnet-core#configure-access-permissions

To use this, you would need to:

a) set up the HttpContextAccessor
b) Create the PrivateAccessControlProvider class
c) Create a controller like this:

``` csharp
[Route("tables/[controller]")]
public class ModelController : TableController<Model>
{
    public ModelController(AppDbContext context, IHttpContextAccessor accessor) : base()
    {
        AccessControlProvider = new PrivateAccessControlProvider<Model>(accessor);
        Repository = new EntityTableRepository<Model>(context);
    }
}
```

The key is the PrivateAccessControlProvider - the IAccessControlProvider is an interface that defines three things:

* A filter that restricts the information that is returned to the client (this is the equivalent of your PerUserFilter)
* A method to determine if the requested action is permitted by the client.
* A pre-commit hook to allow you to update the model prior to writing to the database.

That handles more than 95% of all cases.  In the minority of cases where you need something more than that, you can override any of the TableController methods to get really granular control.

**Reply by ideas2appsaustralia** [reply]
well in that case @fileman was right all along!

**Comment by ideas2appsaustralia** [comment]
Hi

yes it says PerUserFilter but it was an extension on the table 
controller that allowed you to build an odata query on the server side 
if I recall accurately. It came from Adrians iBook using the old 
windows.mobile framework.

anyway regarding the simple query the reason I was unsure about this is 
that it says that with first query it occurs on server then second query 
with the x.Title example occurs on CLient so I understood this to mean 
the whole table is first sent to the mobile client then I query on the 
other fields so still not what I am after.

but maybe I am wrong

scott

On 7/07/2022 6:01 pm, Emanuele Filardo wrote:
>
> your first messagge is about userfilter ... to filter table based on 
> property, you have to use simple filter 
> <https://docs.microsoft.com/en-us/azure/developer/mobile-apps/azure-mobile-apps/howto/client/dotnet#filtering-data>
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/Azure/azure-mobile-apps/discussions/426#discussioncomment-3098256>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AXS3BMQGXVL6UHJ7RS27MJLVS2FD3ANCNFSM52PXMZ4Q>.
> You are receiving this because you authored the thread.Message ID: 
> ***@***.***>
>

-- 
Regards Scott

**Reply by adrianhall** [reply]
So, there are three places you can filter.

1. On the server side, inside an access control provider - used to limit the records that a user can see.
2. On the server side, but instantiated from the client - used to limit which records are sent to the client.
3. On the client side, used for when you are doing a LINQ query that can't be done via OData.

If you do the following:

``` csharp
var records = table.Where(m => m.Color == "Blue").ToListAsync();
```

then the Where clause is done on the server - it's before the ToListAsync().  If you do the following:

``` csharp
var records = (await table.ToListAsync()).Where(m => m.Color == "Blue");
```

The where clause is done on the client side.  You normally will only see the client side stuff when (a) you are doing a calculation (e.g. if you have a latitude and longitude, you combine them into a geographic point, then calculate the distance to the point, then use where to limit the geographic point), or (b) you are doing a join between two tables (it ends up being two full lists and you do the LINQ join between them on the client).

Of course, when operating offline, everything is on the client, but I hope that was obvious.
